### PR TITLE
[pulsar-flink] Implements a batch program on Pulsar topic by writing Flink DataSet as Avro

### PR DESCRIPTION
**Pull request by ambition119**
Implements a batch program on Pulsar topic by writing Flink DataSet as Avro data Type

### CODE REVIEW

**Comment by ambition119** [comment]
```java
[ERROR] testFunctionAssignmentsWithRestart(org.apache.pulsar.functions.worker.PulsarWorkerAssignmentTest)  Time elapsed: 20.029 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.apache.pulsar.functions.worker.PulsarWorkerAssignmentTest.testFunctionAssignmentsWithRestart() didn't finish within the time-out 20000
```

 I commit code couldn't lead to  this error, can you review @merlimat  @JGailor @ivankelly ? tks

**Comment by ambition119** [comment]
```java
2018-12-17\T\11:44:14.874 [ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 87.172 s <<< FAILURE! - in org.apache.pulsar.proxy.server.ProxyLookupThrottlingTest
2018-12-17\T\11:44:14.874 [ERROR] testLookup(org.apache.pulsar.proxy.server.ProxyLookupThrottlingTest)  Time elapsed: 0.025 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
 ```
Is the network slower occur this error?

File: pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        DatumWriter<T> writer = new SpecificDatumWriter<>(t.getSchema());
+        // Output stream to serialize records into byte array.
+        ByteArrayOutputStream arrayOutputStream = new ByteArrayOutputStream();
+        // Low-level class for serialization of Avro values.
+        Encoder encoder = EncoderFactory.get().binaryEncoder(arrayOutputStream, null);
+        arrayOutputStream.reset();
+        try {
+            writer.write(t,encoder);
+            encoder.flush();
+        }catch (IOException e){
```
Comment (sijie) [inline pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java]:
```suggestion
        } catch (IOException e) {
```

File: pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Avro Serialization Schema to serialize Dataset records to Avro.
+ */
+public class AvroSerializationSchema<T extends SpecificRecord> implements SerializationSchema<T> {
+
+    private static final long serialVersionUID = -6691140169413760919L;
+
+    @Override
+    public byte[] serialize(T t) {
+        if (null == t){
```
Comment (sijie) [inline pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java]:
```suggestion
        if (null == t) {
```

File: pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.avro.specific.SpecificRecord;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+/**
+ * Avro Serialization Schema to serialize Dataset records to Avro.
+ */
+public class AvroSerializationSchema<T extends SpecificRecord> implements SerializationSchema<T> {
```
Comment (sijie) [inline pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java]: can you write a simple unit test for this class?

File: pulsar-flink/src/test/java/org/apache/flink/batch/connectors/pulsar/example/FlinkPulsarBatchAvroSinkExample.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/**
```
Comment (sijie) [inline pulsar-flink/src/test/java/org/apache/flink/batch/connectors/pulsar/example/FlinkPulsarBatchAvroSinkExample.java]: can you rebase to latest master? in latest master, all examples are put under `examples` directory.

https://github.com/apache/pulsar/tree/master/examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example

Comment (ambition119) [reply pulsar-flink/src/test/java/org/apache/flink/batch/connectors/pulsar/example/FlinkPulsarBatchAvroSinkExample.java]: I fork pulsar code two days ago, now I add this content. thk

File: pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java (hunk 3 above)
Comment (ambition119) [reply pulsar-flink/src/main/java/org/apache/flink/batch/connectors/pulsar/serialization/AvroSerializationSchema.java]: add AvroSerializationSchemaTest and AvroSerializationSchemaTest

File: examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md [hunk 5]
```diff
@@ -279,5 +279,67 @@ Please find sample output for above application as follows:
 
 ### Complete Example
 
-You can find a complete example [here](https://github.com/apache/incubator-pulsar/tree/master/pulsar-flink/src/test/scala/org/apache/flink/batch/connectors/pulsar/example/FlinkPulsarBatchJsonSinkScalaExample.scala).
+You can find a complete example [here](https://github.com/apache/pulsar/tree/master/examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/FlinkPulsarBatchJsonSinkScalaExample.scala).
 In this example, Flink DataSet is processed and written to Pulsar in Json format.
+
+
+# PulsarJsonOutputFormat
```
Comment (erenavsarogullari) [inline examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md]: @ambition119 Firstly, thanks for the contribution.
Following documentation should point `PulsarAvroOutputFormat` documentation.
Also row: **289** and **304**.

File: examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md [hunk 6]
```diff
@@ -279,5 +279,67 @@ Please find sample output for above application as follows:
...
+      
+          // set parallelism to write Pulsar in parallel (optional)
+          env.setParallelism(2)
+      
+          // execute program
+          env.execute("Flink - Pulsar Batch Avro")
+        }
+```
+
+**Note:** Property definitions of the model should cover `@BeanProperty` to be visible.
```
Comment (erenavsarogullari) [inline examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md]: `@BeanProperty` is required for **Json** format to be visible properties for serialization. So, this note needs to be removed for Avro.

File: examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md (hunk 5 above)
Comment (ambition119) [reply examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md]: ok

File: examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md (hunk 6 above)
Comment (ambition119) [reply examples/flink-consumer-source/src/main/scala/org/apache/flink/batch/connectors/pulsar/example/README.md]: ok

**Pull request merged**